### ISSUE TITLE
Lifetime issue when `#[pin_project]` is used

### ISSUE THREAD

**Issue by Diggsey** [taiki-e/pin-project#185]
[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=44cfef1009b6d5fa949409383532d09e)

This code:
```rust
use pin_project::pin_project;

trait Bar<X> {
    type Y;
}

struct Example<A>(A);

impl<X, T> Bar<X> for Example<T> {
    type Y = Option<T>;
}

#[pin_project]
struct Foo<A, B> {
    x: <Example<A> as Bar<B>>::Y
}
```

Fails to compile with the error:
```
error[E0309]: the parameter type `A` may not live long enough
  --> src/main.rs:13:1
   |
13 | #[pin_project]
   | ^^^^^^^^^^^^^^
14 | struct Foo<A, B> {
   |            - help: consider adding an explicit lifetime bound...: `A: 'pin`
   |
note: ...so that the reference type `&'pin mut std::option::Option<A>` does not outlive the data it points at
  --> src/main.rs:13:1
   |
13 | #[pin_project]
   | ^^^^^^^^^^^^^^
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

It's possibly a compiler bug, because if I change `Option<A>` to `A` it compiles fine, which is really weird...

This is the expanded code produced by pin-project:
```rust
#![feature(prelude_import)]
#[prelude_import]
use std::prelude::v1::*;
#[macro_use]
extern crate std;
use pin_project::pin_project;
trait Bar<X> {
    type Y;
}
struct Example<A>(A);
impl<X, T> Bar<X> for Example<T> {
    type Y = Option<T>;
}
struct Foo<A, B> {
    x: <Example<A> as Bar<B>>::Y,
}
#[allow(clippy::mut_mut)]
#[allow(dead_code)]
struct __FooProjection<'pin, A, B> {
    x: &'pin mut (<Example<A> as Bar<B>>::Y),
}
#[allow(dead_code)]
struct __FooProjectionRef<'pin, A, B> {
    x: &'pin (<Example<A> as Bar<B>>::Y),
}
impl<A, B> Foo<A, B> {
    fn project<'pin>(self: ::core::pin::Pin<&'pin mut Self>) -> __FooProjection<'pin, A, B> {
        unsafe {
            let Foo { x } = self.get_unchecked_mut();
            __FooProjection { x }
        }
    }
    fn project_ref<'pin>(self: ::core::pin::Pin<&'pin Self>) -> __FooProjectionRef<'pin, A, B> {
        unsafe {
            let Foo { x } = self.get_ref();
            __FooProjectionRef { x }
        }
    }
}
#[allow(single_use_lifetimes)]
#[allow(non_snake_case)]
#[deny(safe_packed_borrows)]
fn __pin_project_assert_not_repr_packed_Foo<A, B>(val: &Foo<A, B>) {
    &val.x;
}
#[allow(non_snake_case)]
fn __unpin_scope_Foo() {
    struct __Foo<'pin, A, B> {
        __pin_project_use_generics: ::pin_project::__private::AlwaysUnpin<'pin, (A, B)>,
    }
    impl<'pin, A, B> ::core::marker::Unpin for Foo<A, B> where __Foo<'pin, A, B>: ::core::marker::Unpin {}
}
trait FooMustNotImplDrop {}
#[allow(clippy::drop_bounds)]
impl<T: ::core::ops::Drop> FooMustNotImplDrop for T {}
#[allow(single_use_lifetimes)]
impl<A, B> FooMustNotImplDrop for Foo<A, B> {}
#[allow(single_use_lifetimes)]
impl<A, B> ::pin_project::__private::PinnedDrop for Foo<A, B> {
    unsafe fn drop(self: ::core::pin::Pin<&mut Self>) {}
}
```

**Comment by Diggsey** [comment]
And here is the result of reducing that expanded code:
```rust

trait Bar<X> {
    type Y;
}
struct Example<A>(A);
impl<X, T> Bar<X> for Example<T> {
    type Y = Option<T>;
}
struct Foo<A, B> {
    x: <Example<A> as Bar<B>>::Y,
}
struct __FooProjection<'pin, A, B> {
    x: &'pin mut <Example<A> as Bar<B>>::Y,
}
```

**Comment by Aaron1011** [comment]
It looks like the compiler is giving the correct error here. From the [`infer_outlives` rfc](https://rust-lang.github.io/rfcs/2093-infer-outlives.html#where-explicit-annotations-would-still-be-required):


> In turn, if this associated type were used in a struct, where-clauses would be required. As we'll see in the reference-level explanation, this is a consequence of the fact that we do inference without regard for associated type normalization, but it makes for a relatively simple rule -- explicit where clauses are needed in the preseence of impls like the one above:

```rust
struct Foo<'a, T>
  where T: 'a // still required, not inferred from `field`
{
  field: <Vec<T> as MakeRef<'a>>::Type
}    
```

I think the solution is for `pin-project` to always generate `T: 'pin` for each type parameter `T`.

**Comment by Diggsey** [comment]
@Aaron1011 it might be safer to add this bound instead:
` where Foo<A, B>: 'pin`

This also seems to fix the problem.

**Comment by Aaron1011** [comment]
@Diggsey: That's a good idea - I'll do that instead.

**Comment by taiki-e** [comment]
Published 0.4.9 which fixes this issue.

**Comment by Diggsey** [comment]
Nice! 👍
